File: 2024/day_8/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type Point struct {
	row, col int
}

func (p Point) is_inside_grid(grid []string) bool {
	return p.row >= 0 && p.col >= 0 && p.row < len(grid) && p.col < len(grid[0])
}

func read() []string {
	file, err := os.Open("input.txt")
	if err != nil {
		fmt.Println("Error opening file:", err)
		return nil
	}
	defer file.Close()

	var grid []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		grid = append(grid, line)
	}

	return grid
}

func get_antennas(grid []string) map[rune][]Point {
	antennas := make(map[rune][]Point)
	for row, line := range grid {
		for col, s := range line {
			if s != '.' {
				value, ok := antennas[s]
				if ok {
					value = append(value, Point{row, col})
					antennas[s] = value
				} else {
					antennas[s] = []Point{{row, col}}
				}
			}
		}
	}
	return antennas
}

func get_antinodes(grid []string, ref Point, other Point) <-chan Point {
	ch := make(chan Point)

	go func() {
		defer close(ch)
		drow := ref.row - other.row
		dcol := ref.col - other.col

		i := 1
		for {
			an := Point{ref.row + drow*i, ref.col + dcol*i}
			if an.is_inside_grid(grid) {
				ch <- an // Send the value to the channel
			} else {
				break
			}
			i++
		}
	}()

	return ch
}

func main() {
	grid := read()

	antennas := get_antennas(grid)
	antinodes := make(map[Point]struct{})

	for antenna_type := range antennas {
		antenna_list := antennas[antenna_type]
		for _, ref := range antenna_list {
			for _, other := range antenna_list {
				if ref != other {
					for an := range get_antinodes(grid, ref, other) { // Range over the channel
						antinodes[an] = struct{}{}
						break
					}
				}
			}
		}
	}
	fmt.Printf("First Part: Antinodes: %d\n", len(antinodes))
	fmt.Println("----------------------------------------------")

	for antenna_type := range antennas {
		antenna_list := antennas[antenna_type]
		for _, ref := range antenna_list {
			antinodes[ref] = struct{}{}
			for _, other := range antenna_list {
				antinodes[other] = struct{}{}
				if ref != other {
					for an := range get_antinodes(grid, ref, other) { // Range over the channel
						antinodes[an] = struct{}{}
					}
				}
			}
		}
	}
	fmt.Printf("Second Part: Antinodes: %d\n", len(antinodes))
}

File: 2024/day_8/solution.py
from typing import Tuple
from rich import print
from rich.panel import Panel
from rich.text import Text

def read_input():
    with open("input.txt") as f:
        text = f.read()
        grid = text.splitlines()
        return grid

def find_nearest_antinode(reference: Tuple[int, int], other: Tuple[int, int]):
    drow = reference[0] - other[0]
    dcol = reference[1] - other[1]
    return (reference[0] + drow, reference[1] + dcol)

def antinode_line(reference: Tuple[int, int], other: Tuple[int, int], grid):
    drow = reference[0] - other[0]
    dcol = reference[1] - other[1]

    an_list = [reference, other]
    an = (reference[0] + drow, reference[1] + dcol)
    while (0 <= an[0] < len(grid[0]) and 0 <= an[1] < len(grid)):
        an_list.append(an)
        an = (an[0] + drow, an[1] + dcol)

    return an_list


def generate_panel_content(grid, positions):
    # time.sleep(0.1)
    grid_copy = [list(row) for row in grid]
    for position in positions:
        row, col = position
        elem = grid_copy[row][col]
        if elem == '.':
            grid_copy[row][col] = f"[bold red]#[/bold red]"
        else:
            grid_copy[row][col] = f"[bold red]{elem}[/bold red]"

    s = ''
    for row in grid_copy:
        s += ''.join(row) + '\n'
    return Panel.fit(
        s,
        title=Text(f"Grid - Unique Antinodes {len(positions)}",style="bold blue"),
        border_style="blue",
        padding=(0, 0),  # Remove padding (top/bottom, left/right)
    )

if __name__ == "__main__":
    part2 = 0

    grid = read_input()
    antennas = dict()

    # pylint: disable=C0200
    for row in range(len(grid)):
        for col in range(len(grid[0])):
            elem = grid[row][col]
            if elem not in '.':
                current = antennas.get(elem, [])
                antennas[elem] = current + [(row, col)]
    # print(antennas)
    antinodes = set()

    for ant in antennas:
        to_check =  antennas[ant]
        for ref in to_check:
            for pos in to_check:
                if ref != pos:
                    antinode = find_nearest_antinode(ref, pos)
                    if (0 <= antinode[0] < len(grid[0]) and 0 <= antinode[1] < len(grid)):
                        antinodes.add(antinode)

    print(generate_panel_content(grid, antinodes))
    part1 = len(antinodes)
    print(f"Result of part1: {part1}")

    print("---------------------------")
    for ant in antennas:
        to_check =  antennas[ant]
        for ref in to_check:
            for pos in to_check:
                if ref != pos:
                    an_line = antinode_line(ref, pos, grid)
                    for an in an_line:
                        antinodes.add(an)
    print(generate_panel_content(grid, antinodes))
    part2 = len(antinodes)
    print(f"Result of part2: {part2}")
